### ISSUE TITLE
IndexOption unique:true causing unhandledPromiseRejection MSSQL

### ISSUE THREAD

**Issue by DavidPMacNeil** [typeorm/typeorm#721]
Hello, If I add or modify an index using TypeORM alpha 31, and MSSQL to use `{ unique: true}`, I will get an unhandled promise rejection warning. 

So if I have:
```
@Entity()
@Index('IDX_UNIQUE_CATS', (cat: Cat) => [cat.CatID, cat.Sequence], { unique: true })
class Cat {
    ...
}
```

I will receive this error:
```
(node:3764) UnhandledPromiseRejectionWarning: UnhandledPromiseRejection (rejection id: 1): RequestError: The operation failed because an index or statistics with name 'IDX_UNIQUE_CATS' already exists on table 'Cat'.
```

Thanks for the time.

**Comment by avbentem** [comment]
Same with PostgreSQL, which would throw something like `error: relation "uk_my_column_name" already exists`. 

typeorm 0.1.0-alpha.34 along with pg 7.1.2 towards PostgreSQL 9.6.3 on x86_64-pc-linux-gnu, compiled by gcc (Debian 4.9.2-10) 4.9.2, 64-bit

**Comment by MaikelH** [comment]
Are both of you using the `autoSchemaSync` option?

**Comment by avbentem** [comment]
Yes, I am using `autoSchemaSync: true`, which otherwise works great. (Like everything in TypeORM works great!) And I am sorry for not seeing this earlier:

1. All is fine when creating a new table with a column which has `unique: true` specified for the very first run. For that, the following is used in the `CREATE TABLE` statement:

    ```sql
    CONSTRAINT "uk_my_table_2_my_column_name" UNIQUE ("my_column_name")
    ```

2. However, when **adding `unique: true` to an existing column**, TypeORM adds a constraint with a name that lacks the table name. The `ALTER TABLE ... ADD` uses:

    ```sql
    CONSTRAINT "uk_my_column_name" UNIQUE ("my_column_name")
    ```
    This works fine initially, but fails for subsequent restarts for which TypeORM tries to add the very same constraint again. So, this shorter name seems to be the culprit.

3. And when **adding a column to an existing table and immediately specifying `unique: true` for that new column**, then the `UNIQUE` constraint is not added immediately. Instead, TypeORM only adds the constraint on the next restart, just like above, so then also fails for subsequent restarts.

I guess I could create a PR for (2), which would also partly fix (3), but the latter should obviously add the new constraint in the first `ALTER TABLE ... ADD COLUMN`.

More details below, but I guess the points above cover it all.


## Adding `unique: true` to existing column

With:

```typescript
import {Column, PrimaryGeneratedColumn} from 'typeorm';
import {Entity} from 'typeorm/decorator/entity/Entity';

@Entity('my_table_2')
export class MyTable2 {
    @PrimaryGeneratedColumn()
    id: number;

    @Column({name: 'my_column_name'})
    myColumn: string;
}
```
...the table gets created as:

```sql
executing query: START TRANSACTION
executing query: CREATE SCHEMA IF NOT EXISTS "public"
creating a new table: my_table_2
executing query: CREATE TABLE "my_table_2" ("id" SERIAL NOT NULL, "my_column_name" character varying NOT NULL, PRIMARY KEY("id"))
executing query: COMMIT
```

Next, when changing to

```typescript
    @Column({name: 'my_column_name', unique: true})
```

...then with `logging: 'all'` I see:

```sql
executing query: SELECT * FROM information_schema.tables WHERE table_catalog = 'postgres' AND table_schema = 'public' AND table_name IN ('my_table_1', 'my_table_2')
executing query: SELECT * FROM information_schema.columns WHERE table_catalog = 'postgres' AND table_schema = 'public'
executing query: SELECT t.relname AS table_name, i.relname AS index_name, a.attname AS column_name  FROM pg_class t, pg_class i, pg_index ix, pg_attribute a, pg_namespace ns
WHERE t.oid = ix.indrelid AND i.oid = ix.indexrelid AND a.attrelid = t.oid
AND a.attnum = ANY(ix.indkey) AND t.relkind = 'r' AND t.relname IN ('my_table_1', 'my_table_2') AND t.relnamespace = ns.OID AND ns.nspname ='public' ORDER BY t.relname, i.relname
executing query: SELECT table_name, constraint_name FROM information_schema.table_constraints WHERE table_catalog = 'postgres' AND table_schema = 'public' AND constraint_type = 'FOREIGN KEY'
executing query: SELECT * FROM information_schema.table_constraints WHERE table_catalog = 'postgres' AND table_schema = 'public' AND constraint_type = 'UNIQUE'
executing query: SELECT c.column_name, tc.table_name, tc.constraint_name FROM information_schema.table_constraints tc
JOIN information_schema.constraint_column_usage AS ccu USING (constraint_schema, constraint_name)
JOIN information_schema.columns AS c ON c.table_schema = tc.constraint_schema AND tc.table_name = c.table_name AND ccu.column_name = c.column_name
where constraint_type = 'PRIMARY KEY' AND c.table_schema = 'public' and tc.table_catalog = 'postgres'

executing query: START TRANSACTION
executing query: CREATE SCHEMA IF NOT EXISTS "public"
columns changed in my_table_2. updating: my_column_name
executing query: ALTER TABLE "my_table_2" ADD CONSTRAINT "uk_my_column_name" UNIQUE ("my_column_name")
executing query: COMMIT
```

So, so far so good, the column constraint was added.

However, on subsequent restarts, without any further changes to the model, it shows the same initial queries, and then tries to add the same constraint again:

```sql
executing query: START TRANSACTION
executing query: CREATE SCHEMA IF NOT EXISTS "public"
columns changed in my_table_2. updating: my_column_name
executing query: ALTER TABLE "my_table_2" ADD CONSTRAINT "uk_my_column_name" UNIQUE ("my_column_name")
query failed: ALTER TABLE "my_table_2" ADD CONSTRAINT "uk_my_column_name" UNIQUE ("my_column_name")
error: { error: relation "uk_my_column_name" already exists
    at Connection.parseE (/Users/arjan/dev/workspaces/ipp/server/node_modules/pg/lib/connection.js:546:11)
    at Connection.parseMessage (/Users/arjan/dev/workspaces/ipp/server/node_modules/pg/lib/connection.js:371:19)
    at Socket.<anonymous> (/Users/arjan/dev/workspaces/ipp/server/node_modules/pg/lib/connection.js:114:22)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:548:20)
  name: 'error',
  length: 99,
  severity: 'ERROR',
  code: '42P07',
  detail: undefined,
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'index.c',
  line: '793',
  routine: 'index_create' }
executing query: ROLLBACK
```

Note that `SELECT * FROM information_schema.table_constraints WHERE table_catalog = 'postgres' AND table_schema = 'public' AND constraint_type = 'UNIQUE'` nicely shows the existing `uk_my_column_name` for that very table.


## Adding `unique: true` when creating a new table

When adding the `unique: true` from the beginning, the table gets to be created using:

```sql
executing query: CREATE TABLE "my_table_2" ("id" SERIAL NOT NULL, "my_column_name" character varying NOT NULL, CONSTRAINT "uk_my_table_2_my_column_name" UNIQUE ("my_column_name"), PRIMARY KEY("id"))
```

## Adding new column with `unique: true` to existing table

When adding a new column including `unique: true`, the column is added without the constraint, using:

```sql
executing query: ALTER TABLE "my_table_2" ADD "my_column_name_2" character varying NOT NULL
```

**Comment by DavidPMacNeil** [comment]
@MaikelH Yes I am using `autoSchamSync`.

**Comment by pleerock** [comment]
@iz-iznogood since you are working on indices can you please take a look on this issue btw?

**Comment by iz-iznogood** [comment]
Hi all,

this should be fixed now. The reason for this is because the metadata query for the indices in MSSQL was wrongly excluding the unique indexes and unique constraints, so in every run the system was trying to recreate it.
After my pull requests are merged this should be fixed, I have tested it in several cases.

Issues #719 and #770 are closely related to this issue, I will take a look on them too...

**Comment by pleerock** [comment]
@iz-iznogood thank you!

**Comment by iz-iznogood** [comment]
Pull request #843 has tests for this case. Please check them and close the issue as resolved if it works for you

**Comment by pleerock** [comment]
Fix provided by @iz-iznogood is released in `0.1.0-alpha.40`, please check a new version.
